issue-1010 added ApiKeyResolver to DefaultClientBuilder, DefaultClien…

iancorcoran opened the pull request:
…t and DefaultDataStore

mraible commented on extensions/httpclient/src/test/groovy/com/stormpath/sdk/impl/client/DefaultClientTest.groovy:
```diff
@@ -90,22 +93,40 @@ class DefaultClientTest {
+        def proxy = createMock(com.stormpath.sdk.client.Proxy)
+        def cacheManager = createMock(CacheManager)
+        def authcScheme = AuthenticationScheme.SAUTHC1
+        def clientCredentials = createNiceMock(ClientCredentials)
```
Is there a particular reason you're using a nice mock here? We generally use strict mocks so we can verify the order methods are called on the mock.

mraible commented on extensions/httpclient/src/test/groovy/com/stormpath/sdk/impl/client/DefaultClientTest.groovy:
```diff
@@ -90,22 +93,40 @@ class DefaultClientTest {
     void testGetDataStore() {
 
         def apiKeyCredentials = createNiceMock(ApiKeyCredentials)
+        def apiKeyResolver = createNiceMock(ApiKeyResolver)
```
strict mock

mraible commented on extensions/httpclient/src/test/groovy/com/stormpath/sdk/impl/client/DefaultClientTest.groovy:
```diff
@@ -115,6 +136,7 @@ class DefaultClientTest {
     @Test
     void testDirtyPropertiesNotSharedAmongDifferentResourcesWithSameHref() {
         def apiKeyCredentials = createNiceMock(ApiKeyCredentials)
+        def apiKeyResolver = createNiceMock(ApiKeyResolver)
```
Strict mock preferred

mraible commented on extensions/spring/boot/stormpath-webmvc-spring-boot-starter/src/main/java/com/stormpath/spring/boot/autoconfigure/StormpathWebMvcAutoConfiguration.java:
```diff
@@ -423,6 +423,19 @@ public LocaleChangeInterceptor stormpathLocaleChangeInterceptor() {
+        return super.stormpathRequestApplicationAttributeNames();
+    }
+
+
```
Remove extra line break.

mraible commented on extensions/spring/boot/stormpath-webmvc-spring-boot-starter/src/main/java/com/stormpath/spring/boot/autoconfigure/StormpathWebMvcAutoConfiguration.java:
```diff
@@ -423,6 +423,19 @@ public LocaleChangeInterceptor stormpathLocaleChangeInterceptor() {
     }
 
     @Bean
+    @ConditionalOnMissingBean(name = "stormpathRequestClientAttributeNames")
```
Are these new beans related to this PR?

iancorcoran replied:
These were part of an earlier PR of mine that went in with 1.0.4. Looks like @jarias may have accidentally overwritten them when merging the 1.1 branch into master. Not strictly related to adding an ApiKeyResolver I guess.

mraible commented on impl/src/main/java/com/stormpath/sdk/impl/ds/DefaultDataStore.java:
```diff
@@ -107,37 +107,38 @@
 
-    public DefaultDataStore(RequestExecutor requestExecutor, ApiKeyCredentials apiKeyCredentials) {
-        this(requestExecutor, DEFAULT_API_VERSION, apiKeyCredentials);
+    public DefaultDataStore(RequestExecutor requestExecutor, ApiKeyCredentials apiKeyCredentials, ApiKeyResolver apiKeyResolver) {
```
Add `since` javadoc to these new constructors.

iancorcoran replied:
I don't mind adding these but as a rule this doesn't seem to be done in the impl module outside of class level javadoc?

mraible commented on impl/src/main/java/com/stormpath/sdk/impl/client/DefaultClientBuilder.java:
```diff
@@ -242,6 +244,12 @@ public ClientBuilder setClientCredentials(ClientCredentials clientCredentials) {
         return this;
     }
 
+    public ClientBuilder setApiKeyResolver(ApiKeyResolver apiKeyResolver) {
```
Add `since` javadoc.

mraible commented on impl/src/main/java/com/stormpath/sdk/impl/config/ClientConfiguration.java:
```diff
@@ -58,6 +60,14 @@ public void setApiKeySecret(String apiKeySecret) {
         this.apiKeySecret = apiKeySecret;
     }
 
+    public ApiKeyResolver getApiKeyResolver() {
```
Add `since` javadoc to new methods.

mraible commented on impl/src/test/groovy/com/stormpath/sdk/impl/client/DefaultClientBuilderTest.groovy:
```diff
@@ -88,4 +91,68 @@ class DefaultClientBuilderTestCustomCredentials{
     }
 
+    @Test
+    void testCustomClientCredentialsRequireApiKeyResolver(){
```
We sometimes add `since` javadoc to new tests. Don't know if it's necessary, but doesn't hurt to be thorough!

mraible commented on impl/src/test/groovy/com/stormpath/sdk/impl/client/DefaultClientBuilderTest.groovy:
```diff
@@ -88,4 +91,68 @@ class DefaultClientBuilderTestCustomCredentials{
+        assertEquals testClient.dataStore.apiKey.secret, keySecret
+    }
+
+
```
Remove extra line breaks.

mraible commented on impl/src/test/groovy/com/stormpath/sdk/impl/client/DefaultClientTest.groovy:
```diff
@@ -27,6 +28,7 @@ class DefaultClientTest extends PowerMockTestCase {
     void testCreateRequestExecutor() {
 
         def apiKeyCredentials = createMock(ApiKeyCredentials)
+        def apiKeyResolver = createNiceMock(ApiKeyResolver)
```
strict mock

mraible commented on impl/src/test/groovy/com/stormpath/sdk/impl/directory/DefaultCustomDataTest.groovy:
```diff
@@ -218,7 +219,7 @@ class DefaultDataStoreDelegateTo extends DefaultDataStore {
 
     DefaultDataStoreDelegateTo(RequestExecutor requestExecutor, CustomData customData) {
-        super(requestExecutor, "https://api.stormpath.com/v1", (ApiKeyCredentials)createNiceMock(ApiKeyCredentials))
+        super(requestExecutor, "https://api.stormpath.com/v1", (ApiKeyCredentials)createNiceMock(ApiKeyCredentials), createNiceMock(ApiKeyResolver))
```
I'd suggest `createStrictMock`, but I guess nice mock is used for `ApiKeyCredentials`, so I'd simply suggest consistency - which you have here.

mraible commented on impl/src/test/groovy/com/stormpath/sdk/impl/ds/DefaultDataStoreTest.groovy:
```diff
@@ -390,4 +403,20 @@ class DefaultDataStoreTest {
+    void testApiKeyResolverReturnsCorrectApiKey(){
+        def requestExecutor = createNiceMock(RequestExecutor)
+        def apiKeyForCredentials = createNiceMock(ApiKey)
+        def apiKeyForResolver = createNiceMock(ApiKey)
```
I'd suggest strict over nice mocks.

mraible commented:
Reviewed - back to you @iancorcoran.

iancorcoran commented:
@mraible Believe I've addressed your comments outside of the open javadoc question. In regards to your question around whether the DefaultApiKeyResolver should be added as a property that can be overridden... we have an internal use case where we need to supply our own ApiKeyResolver (ping me on slack if you want to discuss more on the details of it). Similar to supplying custom ClientCredentials to the ClientBuilder, it is not really applicable outside of our internal use case, and follows the same pattern of just being exposed on the DefaultClientBuilder rather than the api ClientBuilder interface.
